facebookresearch/ParlAI
# [FSDP] Zero 3 Optimization Support

# Patch description
This PR adds support for zero3 fsdp optimization. Specify via `--ddp-backend zero3`. Zero3 optimization shards not only the optimizer and gradients but also the model weights. This improves memory pressure, and is especially useful for larger models. Note that this can increase latency as there is more communication cost, and as such for smaller models there may be a slight hit to speed compared to zero2 (and model parallel, of course).

Addresses #3753

**NOTE: This requires pytorch >= 1.12 for use**

# Testing steps
- Enabled the zero3 tests already in CI (thanks Stephen!)
- Confirmed that zero3 works when the number of validation examples is not divisible by the number of GPUs
- Confirmed that zero3 works when `--skip-generation False` during training. 
- Conducted a comprehensive analysis of a variety of "staple" models within ParlAI under different distributed settings; see screenshots below

The main conclusions from each of the models below are:
- **Performance** remains roughly the same regardless of model parallel, zero2, or zero3
- **Speed**: for BART, zero2 remains faster. For the other models, we see varying results; ultimately, everything is faster than model parallel
- **GPU Memory**: zero3 

## BART-Large (400M)
<img width="1076" alt="Screen Shot 2022-12-01 at 2 53 15 PM" src="https://user-images.githubusercontent.com/8562788/205167151-6f3cfd60-185e-4cfb-8e99-16108c3ba1da.png">

## T5-Large (770M)
<img width="1074" alt="Screen Shot 2022-12-01 at 2 51 32 PM" src="https://user-images.githubusercontent.com/8562788/205167233-259d4c0c-6845-4933-9b24-9e18fcff0175.png">

## Reddit 2.7B (base of BlenderBot) 
<img width="1074" alt="Screen Shot 2022-12-01 at 2 51 51 PM" src="https://user-images.githubusercontent.com/8562788/205167299-44a7f86b-3702-4917-98bb-2b2d4d2faef6.png">

## GPT2-XL (1.5B)
<img width="1074" alt="Screen Shot 2022-12-01 at 2 52 16 PM" src="https://user-images.githubusercontent.com/8562788/205167465-567eb097-8312-4fb1-b197-ae02b5fa7571.png">

## R2C2 2.7B
<img width="1078" alt="Screen Shot 2022-12-01 at 2 52 31 PM" src="https://user-images.githubusercontent.com/8562788/205167505-d580eb0b-8174-4148-a1b3-edb650feea79.png">
